### PR TITLE
add IndicatorFilter support

### PR DESCRIPTION
**Pull request by twicksell**

### CODE REVIEW

File: health-api/src/test/java/com/netflix/runtime/health/api/IndicatorFiltersTest.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+ * limitations under the License.
+ */
+package com.netflix.runtime.health.api;
+
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+import org.junit.Test;
+
+public class IndicatorFiltersTest {
```
Comment (drobertduke) [inline health-api/src/test/java/com/netflix/runtime/health/api/IndicatorFiltersTest.java]: Maybe add a test for an empty IndicatorFilter, like
`assertTrue(IndicatorFilters.build().matches(new A()));`
If that's the correct behavior

File: health-api/src/main/java/com/netflix/runtime/health/api/IndicatorFilters.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    public static IndicatorFilterBuilder includes(String... indicatorNames) {
+        return new IndicatorFilterBuilder().includes(indicatorNames);
+    }
+    
+    public static IndicatorFilterBuilder excludes(String... indicatorNames) {
+        return new IndicatorFilterBuilder().excludes(indicatorNames);
+    }
+    
+    public static IndicatorFilter build() {
+        return i -> true;
```
Comment (jhspaybar) [inline health-api/src/main/java/com/netflix/runtime/health/api/IndicatorFilters.java]: Does Java allow _ as an argument name here to signify unused?

File: health-api/src/main/java/com/netflix/runtime/health/api/IndicatorFilters.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+/**
+ * Static builder of for creating an {@link IndicatorFilter} instance.
+ * 
+ * <pre>
+ *  IndicatorFilters.includes(someIndicatorInstance)
+ *                  .excludes("someIndicatorName")
+ *                  .build();
+ * </pre>
+ */
+public class IndicatorFilters {
```
Comment (jhspaybar) [inline health-api/src/main/java/com/netflix/runtime/health/api/IndicatorFilters.java]: We've typically had a method like newBuilder() or builder() to kick off the chain.

File: health-api/src/main/java/com/netflix/runtime/health/api/IndicatorFilters.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * Static builder of for creating an {@link IndicatorFilter} instance.
+ * 
+ * <pre>
+ *  IndicatorFilters.includes(someIndicatorInstance)
```
Comment (jhspaybar) [inline health-api/src/main/java/com/netflix/runtime/health/api/IndicatorFilters.java]: This implies to me that includes can take an instance of a class, but it seems to me it just takes strings?

Comment (twicksell) [inline health-api/src/main/java/com/netflix/runtime/health/api/IndicatorFilters.java]: Its overloaded to take either. Went through a couple of implementations where I specifically needed one or the other. I ended up in a place where I can use either, but left both methods.

File: health-api/src/main/java/com/netflix/runtime/health/api/IndicatorFilters.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+        public IndicatorFilterBuilder excludes(String... indicatorNames) {
+            if (indicatorNames != null) {
+                excludedIndicatorNames.addAll(Arrays.asList(indicatorNames));
+            }
+            return this;
+        }
+        
+        public IndicatorFilter build() {
+            return indicatorName -> {
+               Optional<Boolean> include = includedIndicatorNames.stream().map(included -> indicatorName.equals(included)).reduce((a,b) -> a || b);
```
Comment (jhspaybar) [inline health-api/src/main/java/com/netflix/runtime/health/api/IndicatorFilters.java]: I think this boils down to essentially `return excluded.stream().allMatch(n -> !indicatorName.equals(i)) && (included.isEmpty() || included.stream().anyMatch(indicatorName::equals));`

If you don't want to overly compress all the if blocks you should at least be able to ditch the map reduce.

File: health-api/src/test/java/com/netflix/runtime/health/api/IndicatorFiltersTest.java [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    public void testAdditiveExclude() {
+        assertFalse(IndicatorFilters.excludes(A.class.getName()).excludes(B.class.getName()).build().matches(new A()));
+        assertFalse(IndicatorFilters.excludes(A.class.getName()).excludes(B.class.getName()).build().matches(new B()));
+        assertTrue(IndicatorFilters.excludes(A.class.getName()).excludes(B.class.getName()).build().matches(new C()));
+    }
+    
+    @Test
+    public void testIncludesAndExcludes() {
+        assertTrue(IndicatorFilters.includes(A.class.getName()).excludes(B.class.getName()).build().matches(new A()));
+        assertFalse(IndicatorFilters.includes(A.class.getName()).excludes(B.class.getName()).build().matches(new B()));
```
Comment (jhspaybar) [inline health-api/src/test/java/com/netflix/runtime/health/api/IndicatorFiltersTest.java]: Doesn't this test still pass if you don't have the excludes at all?

File: health-integrations/README.md [hunk 7]
```diff
@@ -14,3 +15,27 @@ InjectorBuilder
...
+        new ArchaiusHealthStatusFilterModule(),
+        // ... other modules ...
+    )
+    .createInjector();
+```
+
+Indicators can now be included/excluded by setting the following properties in Archaius2.
+```
+health.status.indicators.include=com.myproject.MyFilter
+health.status.indicators.exclude=com.myproject.DontIncludeThisFilter
```
Comment (jhspaybar) [inline health-integrations/README.md]: Isn't there no situation in which you'd specify both include and exclude?  It seems to me if I have just the first line here `health.status.indicators.include=com.myproject.MyFilter` that everything would be excluded that wasn't com.myproject.MyFilter?

File: health-core/src/main/java/com/netflix/runtime/health/core/SimpleHealthCheckAggregator.java [hunk 8]
```diff
@@ -121,31 +128,35 @@ public void run() {
...
-		    			.withDetail(Health.NAME_KEY, getIndicatorName(callback.getIndicator()))
-		    			.build();
-		    	healths.add(health);
-		    	return health;
+		    			.withDetail(Health.NAME_KEY, callback.getIndicator().getName()).build();
+		    	if(callback.isSuppressed()) {
+		    	    suppressedHealths.add(health);
+		    	    return Health.healthy().build();
+		    	}
+		    	else {
```
Comment (jhspaybar) [inline health-core/src/main/java/com/netflix/runtime/health/core/SimpleHealthCheckAggregator.java]: } else { on 1 line please

File: health-integrations/src/main/java/com/netflix/runtime/health/eureka/EurekaHealthStatusBridgeModule.java [hunk 9]
```diff
@@ -33,38 +33,48 @@
...
             implements ApplicationEventListener<HealthCheckStatusChangedEvent>, LifecycleListener {
 
-        @com.google.inject.Inject(optional=true)
+        @com.google.inject.Inject(optional = true)
         private ApplicationEventDispatcher eventDispatcher;
-        @Inject private ApplicationInfoManager applicationInfoManager;
-        @Inject private HealthCheckAggregator healthCheckAggregator;
-        @Inject private EurekaClient eurekaClient;
+        @Inject
+        private ApplicationInfoManager applicationInfoManager;
```
Comment (jhspaybar) [inline health-integrations/src/main/java/com/netflix/runtime/health/eureka/EurekaHealthStatusBridgeModule.java]: If we use constructor or setter injection here it should be easier to test things than with field injection.  If we use constructor injection it should also be possible to ditch the init() PostContruct method in favor of just doing that in a constructor.

File: health-integrations/src/main/java/com/netflix/runtime/health/eureka/EurekaHealthStatusBridgeModule.java [hunk 10]
```diff
@@ -79,14 +89,18 @@ public String getId() {
...
+
         @Override
         public void onStarted() {
             eurekaClient.registerHealthCheck(new HealthCheckHandler() {
                 @Override
                 public InstanceStatus getStatus(InstanceStatus currentStatus) {
                     try {
-                        return getInstanceStatusForHealth(healthCheckAggregator.check().get());
+                        if(filter != null) {
+                            return getInstanceStatusForHealth(healthCheckAggregator.check().get());
```
Comment (jhspaybar) [inline health-integrations/src/main/java/com/netflix/runtime/health/eureka/EurekaHealthStatusBridgeModule.java]: Shouldn't this method pass the filter to check() ?

File: health-integrations/README.md (hunk 7 above)
Comment (twicksell) [inline health-integrations/README.md]: Excludes do take priority if you've included a large set and only want to exclude one for some reason. But in general I expect people to only want to use one. Up to you to decide which way you prefer. Including only your own would give you the behavior of not getting any other indicators if that is what you want.

File: health-integrations/src/main/java/com/netflix/runtime/health/eureka/EurekaHealthStatusBridgeModule.java [hunk 11]
```diff
@@ -33,38 +33,48 @@
 import com.netflix.governator.spi.LifecycleListener;
 import com.netflix.runtime.health.api.HealthCheckAggregator;
 import com.netflix.runtime.health.api.HealthCheckStatus;
+import com.netflix.runtime.health.api.IndicatorFilter;
 import com.netflix.runtime.health.core.HealthCheckStatusChangedEvent;
+import com.netflix.runtime.health.status.ArchaiusHealthStatusFilterModule;
```
Comment (jhspaybar) [inline health-integrations/src/main/java/com/netflix/runtime/health/eureka/EurekaHealthStatusBridgeModule.java]: Is this being used?

File: health-integrations/src/main/java/com/netflix/runtime/health/servlet/HealthStatusServlet.java [hunk 12]
```diff
@@ -26,23 +26,31 @@
 
 import com.netflix.runtime.health.api.HealthCheckAggregator;
 import com.netflix.runtime.health.api.HealthCheckStatus;
+import com.netflix.runtime.health.api.IndicatorFilter;
+import com.netflix.runtime.health.status.ArchaiusHealthStatusFilterModule;
```
Comment (jhspaybar) [inline health-integrations/src/main/java/com/netflix/runtime/health/servlet/HealthStatusServlet.java]: Doesn't look used again.

File: health-integrations/src/main/java/com/netflix/runtime/health/eureka/EurekaHealthStatusBridgeModule.java (hunk 9 above)
Comment (twicksell) [inline health-integrations/src/main/java/com/netflix/runtime/health/eureka/EurekaHealthStatusBridgeModule.java]: I'm stuck doing field injection because of the optional inject. Not aware of any other way to get around this with Guice.

File: health-integrations/src/main/java/com/netflix/runtime/health/status/ArchaiusHealthStatusFilterModule.java [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+
+        public ArchaiusDrivenStatusFilter(HealthStatusInclusionConfiguration config) {
+            this.config = config;
+        }
+
+        @Override
+        public boolean matches(String indicatorName) {
+            return IndicatorFilters
+                    .includes(config.includedIndicators())
+                    .excludes(config.excludedIndicators()).build()
```
Comment (jhspaybar) [inline health-integrations/src/main/java/com/netflix/runtime/health/status/ArchaiusHealthStatusFilterModule.java]: Are we rebuilding on every match() call to pick up fast property changes?  Is there any implication if if this picks up changes halfway through iterating  the list of indicators or anything?

File: health-integrations/src/main/java/com/netflix/runtime/health/status/ArchaiusHealthStatusFilterModule.java [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+        private final HealthStatusInclusionConfiguration config;
+
+        public ArchaiusDrivenStatusFilter(HealthStatusInclusionConfiguration config) {
+            this.config = config;
+        }
+
+        @Override
+        public boolean matches(String indicatorName) {
+            return IndicatorFilters
+                    .includes(config.includedIndicators())
```
Comment (jhspaybar) [inline health-integrations/src/main/java/com/netflix/runtime/health/status/ArchaiusHealthStatusFilterModule.java]: I'm confused how this works, doesn't `.includes` take a varargs of String, but it seems that config.includedIndicators is returning a List<Class<?>>.  I know the List is fine, but it seems the Class<?> is wrong?  Maybe there's some magic here I'm missing or I got myself confused somewhere.

File: health-integrations/src/main/java/com/netflix/runtime/health/servlet/HealthStatusServlet.java (hunk 12 above)
Comment (twicksell) [inline health-integrations/src/main/java/com/netflix/runtime/health/servlet/HealthStatusServlet.java]: Apparently that is from my linking the module in some of the javadoc. Thanks for pointing that out.

File: health-api/src/main/java/com/netflix/runtime/health/api/IndicatorFilters.java (hunk 4 above)
Comment (jhspaybar) [inline health-api/src/main/java/com/netflix/runtime/health/api/IndicatorFilters.java]: I'm just not seeing the overload in this diff, maybe I'm just blind?

File: health-integrations/src/main/java/com/netflix/runtime/health/status/ArchaiusHealthStatusFilterModule.java (hunk 13 above)
Comment (twicksell) [inline health-integrations/src/main/java/com/netflix/runtime/health/status/ArchaiusHealthStatusFilterModule.java]: We are rebuilding every time. Talked to David about using the property change listener events and synchronizing on some internal reference to that property, but it didn't seem to be worth it given the (in)frequency of these calls. As for changes to config in the middle of an operation, once we get past the creation of this filter it isn't an issue. It is possible that includes and excludes could be updated simultaneously but read independently. But there is no api for transactions around config changes. Worst case, we have the update for only one of the two properties for a single iteration of the health check.

File: health-integrations/src/main/java/com/netflix/runtime/health/status/ArchaiusHealthStatusFilterModule.java (hunk 14 above)
Comment (twicksell) [inline health-integrations/src/main/java/com/netflix/runtime/health/status/ArchaiusHealthStatusFilterModule.java]: config returns String[], which is valid for the String... varag

File: health-api/src/main/java/com/netflix/runtime/health/api/IndicatorFilters.java (hunk 4 above)
Comment (twicksell) [inline health-api/src/main/java/com/netflix/runtime/health/api/IndicatorFilters.java]: Check the interface, IndicatorFilter line 25

Comment (jhspaybar) [inline health-api/src/main/java/com/netflix/runtime/health/api/IndicatorFilters.java]: Okay, but the builder doesn't let me use classes?  Is that an oversight?

Comment (twicksell) [inline health-api/src/main/java/com/netflix/runtime/health/api/IndicatorFilters.java]: It does not allow for classes. Not an oversight. There was originally a notion that indicator's should be named, so I kept that around and built the filters around that. To support classes as a parameter I'd have to be able to instantiate the class, which puts constraints on what kind of constructor an indicator could have. Didn't really want that.

File: health-api/src/main/java/com/netflix/runtime/health/api/HealthCheckAggregator.java [hunk 15]
```diff
@@ -18,7 +18,22 @@
 import java.util.concurrent.CompletableFuture;
 
 /**
+ * The HealthCheckAggregator wraps all registered implementations of {@link HealthIndicator} and is responsible
```
Comment (elandau) [inline health-api/src/main/java/com/netflix/runtime/health/api/HealthCheckAggregator.java]: 'all registered implementations' seems to specific.  Consider generalizing this without making assumption about where the HealthIndicators came from.

File: health-api/src/main/java/com/netflix/runtime/health/api/HealthCheckStatus.java [hunk 16]
```diff
@@ -26,26 +26,48 @@
     private final boolean isHealthy;
 
     private final List<Health> healthResults;
+    private final List<Health> suppressedHealthResults;
 
     public HealthCheckStatus(boolean isHealthy, List<Health> indicators) {
         this.isHealthy = isHealthy;
         this.healthResults = new ArrayList<>(indicators);
+        this.suppressedHealthResults = new ArrayList<>();
```
Comment (elandau) [inline health-api/src/main/java/com/netflix/runtime/health/api/HealthCheckStatus.java]: Is it necessary to track this as a separate list?  Can 'suppressed' be an attribute of Health instead?

File: health-api/src/main/java/com/netflix/runtime/health/api/IndicatorFilter.java [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+ * limitations under the License.
+ */
+package com.netflix.runtime.health.api;
+
+/**
+ * A filter used when invoking {@link HealthCheckAggregator}.check() to explicitly include/exclude which
+ * indicators should be used when calculating the healthiness of the application. See {@link IndicatorFilters}
+ * for creating IndicatorFilters programmatically. 
+ */
+public interface IndicatorFilter {
```
Comment (elandau) [inline health-api/src/main/java/com/netflix/runtime/health/api/IndicatorFilter.java]: 'Filter' and 'matches' don't sound right together.  How about calling this class IndicatorMatcher or calling it IndicatorPredicate and change 'matches' to 'apply'.

File: health-api/src/main/java/com/netflix/runtime/health/api/IndicatorFilters.java (hunk 3 above)
Comment (elandau) [inline health-api/src/main/java/com/netflix/runtime/health/api/IndicatorFilters.java]: +1

File: health-api/src/main/java/com/netflix/runtime/health/api/IndicatorFilters.java [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+ * 
+ * <pre>
+ *  IndicatorFilters.includes(someIndicatorInstance)
+ *                  .excludes("someIndicatorName")
+ *                  .build();
+ * </pre>
+ */
+public class IndicatorFilters {
+    
+    public static IndicatorFilterBuilder includes(String... indicatorNames) {
```
Comment (elandau) [inline health-api/src/main/java/com/netflix/runtime/health/api/IndicatorFilters.java]: Indicator names seems to restrictive here.  How about using a Matcher interface like this

```java
interface Matcher {
    boolean matches(HealthIndicator indicator);
}
```
 You can then have different matchers based on classes, packages, indicator names, existence of annotations, etc..

File: health-api/src/main/java/com/netflix/runtime/health/api/IndicatorFilter.java [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+
+    default boolean matches(HealthIndicator indicator) {
+        return matches(indicator.getName());
+    }
+
+    boolean matches(String indicatorName);
```
Comment (elandau) [inline health-api/src/main/java/com/netflix/runtime/health/api/IndicatorFilter.java]: Having two methods here complicates the API.  I'd much rather have the single method above.  Matching by name should be delegated to a ByNameIndicatorFilter implementation instead.

File: health-api/src/main/java/com/netflix/runtime/health/api/IndicatorFilters.java (hunk 18 above)
Comment (elandau) [inline health-api/src/main/java/com/netflix/runtime/health/api/IndicatorFilters.java]: Having different excludes and includes in addition to the IndicatorFilter abstraction is confusing to me.  IndicatorFilter is inclusive so all we need is a 'not' method like this,

```java
public static IndicatorFilter not(IndicatorFilter delegate) {

}
```

File: health-api/src/main/java/com/netflix/runtime/health/api/HealthCheckStatus.java (hunk 16 above)
Comment (twicksell) [inline health-api/src/main/java/com/netflix/runtime/health/api/HealthCheckStatus.java]: I started there, but that ends up propagating this pattern of having to check isHealthy && !isSuppressed for each Indicator which feels wrong.

File: health-api/src/main/java/com/netflix/runtime/health/api/IndicatorFilters.java (hunk 3 above)
Comment (twicksell) [inline health-api/src/main/java/com/netflix/runtime/health/api/IndicatorFilters.java]: We didn't do that in Health. Its just Health.healthy().withDetails(...).build(). IndicatorFilters.builder.includes(...).build() feels pretty redundant. Does this add something I'm missing?

File: health-api/src/main/java/com/netflix/runtime/health/api/IndicatorFilters.java [hunk 20]
```diff
@@ -0,0 +1,86 @@
...
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * Static builder of for creating an {@link IndicatorFilter} instance.
+ * 
+ * <pre>
+ *  IndicatorFilters.includes(someIndicatorInstance)
+ *                  .excludes("someIndicatorName")
```
Comment (qiangdavidliu) [inline health-api/src/main/java/com/netflix/runtime/health/api/IndicatorFilters.java]: (perhaps just documentation) there's some missing explanation on how conflicting includes and excludes conflict resolve here. what happens if I include and exclude the same indicator name? What happens to this specific indicator? What happens implicitly to the rest of the indicators?

File: health-integrations/README.md [hunk 21]
```diff
@@ -14,3 +15,27 @@ InjectorBuilder
...
+        new EurekaHealthStatusBridgeModule(),
+        new ArchaiusHealthStatusFilterModule(),
+        // ... other modules ...
+    )
+    .createInjector();
+```
+
+Indicators can now be included/excluded by setting the following properties in Archaius2. Excluded indicators will take priority over included ones.
+```
+health.status.indicators.include=com.myproject.MyIndicator
```
Comment (qiangdavidliu) [inline health-integrations/README.md]: Let's make this example with multiple included indicators so that how to do this for a list of indicators can be easily understood.

File: health-integrations/build.gradle [hunk 22]
```diff
@@ -5,7 +5,9 @@ plugins {
 dependencies {
     compile project(':health-core')
     compile "com.netflix.eureka:eureka-client:${eurekaVersion}", optional
+    compile "com.netflix.archaius:archaius2-core:${archaiusVersion}"
```
Comment (qiangdavidliu) [inline health-integrations/build.gradle]: Adding archaius2 here means consumers of health-integrations will now be using a mix of archaius1 (from eureka) and archaius2. This is fine, but we should either (or maybe both):
1. add instructions on how to install the bridge module in the README
2. clean up eureka-client-archaius2 and use that here instead.

File: health-api/src/test/java/com/netflix/runtime/health/api/IndicatorMatchersTest.java [hunk 23]
```diff
@@ -0,0 +1,90 @@
...
+    @Test
+    public void testIncludesAndExcludes() {
+        assertTrue(IndicatorMatchers.includes(A.class.getName()).excludes(B.class.getName()).build().matches(new A()));
+        assertFalse(IndicatorMatchers.includes(A.class.getName()).excludes(B.class.getName()).build().matches(new B()));
+        assertFalse(IndicatorMatchers.includes(A.class.getName()).excludes(B.class.getName()).build().matches(new C()));
+    }
+    
+    @Test
+    public void testExcludeBeatsInclude() {
+        assertFalse(IndicatorMatchers.includes(A.class.getName()).excludes(A.class.getName()).build().matches(new A()));
```
Comment (jhspaybar) [inline health-api/src/test/java/com/netflix/runtime/health/api/IndicatorMatchersTest.java]: It might also be helpful to prove there's no order dependency by including A after excluding it.

**Comment by jhspaybar** [comment]
:+1: LGTM
